### PR TITLE
[Xamarin.Android.Build.Tasks] make sure `Disposable ()` is called for assembly

### PR DESCRIPTION
**Pull request by lewurm**
see https://github.com/jbevain/cecil/issues/291

This fixes the `Sharing violation` issue we've seen in the `RemoveRegisterAttribute` task. I also attempted to fix #44529.

Depends on https://github.com/xamarin/java.interop/pull/87, then this PR needs an update.

### CODE REVIEW

**Comment by lewurm** [comment]
This should supersede https://github.com/xamarin/xamarin-android/pull/230. I couldn't repro it though, @radekdoulik could you test it please?

File: external/Java.Interop [hunk 1]
```diff
@@ -1 +1 @@
-Subproject commit a1d3ecc8ba6e67b96ae1c633194b0e78c2ac5c23
+Subproject commit d2b5bc101c6643dac5d377451a7a84d232adf93e
```
Comment (dellis1972) [inline external/Java.Interop]: Not sure we need a java.interop change, you might need to update your submodules'

Comment (lewurm) [inline external/Java.Interop]: there're changes in `Java.Interop` required, see https://github.com/xamarin/java.interop/pull/87.

**Comment by lewurm** [comment]
updated PR, now works with the repro provided by https://bugzilla.xamarin.com/show_bug.cgi?id=44529#c2

However, it's a bit tricky: `DirectoryAssemblyResolver` caches the results when it loads assemblies, so we need to make sure that we already pass the right `ReaderParameters` to `res.Load ()`. The `ReaderParameters` are actually not part of the key in the cache, so they get ignored in `res.GetAssembly ()`. Also, using `using` for `res.GetAssembly ()` would be wrong (we would dispose it, but there's still a reference to it in the cache).

I _think_ `Dispose ()` of `DirectoryAssemblyResolver` should dispose all assemblies in its cache _or_ not cache at all (`Mono.Cecil.BaseAssemblyResolver` doesn't cache either). I'll let someone else decide that.

**Comment by jonpryor** [comment]
> I *think* `Dispose ()` of `DirectoryAssemblyResolver` should dispose all assemblies in its cache *or* not cache at all (`Mono.Cecil.BaseAssemblyResolver` doesn't cache either).

It should dispose; that is part of https://github.com/xamarin/java.interop/pull/88

**Comment by jonpryor** [comment]
This PR is superseded by PR #233.

**Pull request closed without merging**